python/ray/air/util/tensor_extensions/arrow.py

```diff
@@ -270,3 +271,55 @@ def to_numpy(self, zero_copy_only: bool = True):
             A single ndarray representing the entire array of tensors.
         """
         return self._to_numpy(zero_copy_only=zero_copy_only)
+
+
+def _copy_tensor_array_if_needed(a: "ArrowTensorArray"):
+    """Copy tensor array if it's a zero-copy slice. This is to circumvent an Arrow
```

Comment:
A few questions mostly to understand Arrow: How and where did we check if given "a" is zero copy? If it is not zero-copy slice, what do we do/return? Do copied_offset_buf and data_buf actually make a copy (since they are sliced from buffer)?

Reply:
I'll also add a comment to the code describing this since it's a bit subtle but here is the gist:

Arrow has both **array-level slicing** and **buffer-level slicing**; both are zero-copy, but the former has a serialization bug where the entire buffer is serialized instead of just the slice, while the latter's serialization works as expected and only serializes the slice of the buffer. I.e., array-level slicing doesn't propagate the slice down to the buffer when serializing the array.

All that these copy methods do is, at serialization time, take the array-level slicing and translate them to buffer-level slicing, so only the buffer slice is sent over the wire instead of the entire buffer.